docs/plugins/npe2_manifest_specification.md

```diff
@@ -0,0 +1,606 @@
+```
+
+**Parameters**
+
+path(str)
+: Path to file, directory, or resource (like a url).
+
+**Return**
+
+layer_data(list of LayerData)
+: Each `LayerData` element is a tuple of `(data,meta,layer_type)`.
+
+## Writers
+
+### Required fields
+
+- **command** Identifier of the _command_ providing the writer.
+- **layer_types** List of layer type constraints. These determine what
+  combinations of layers this writer handles.
+
+### Optional fields
+
+- **display_name** Brief text used to describe this writer when presented.
+  Empty by default. When present this is presented along side the plugin name
+  and may be used to distinguish the kind of writer for the user. E.g. "lossy"
+  or "lossless".
+- **filename_extensions** List of filename extensions compatible with this
+  writer. The first entry is used as the default if necessary. Empty by default.
+  When empty, any filename extension is accepted.
+
+### Examples
+
+**Example**
+
+Single-layer writer
+
+```yaml
+name: napari
+contributions:
+  commands:
+    - id: napari.write_points
+      python_name: napari.plugins._builtins:napari_write_points
+      title: Save points layer to csv
+  writers:
+    - command: napari.write_points
+      filename_extensions: [".csv"]
+      layer_types: ["points"]
+```
+
+**Example**
+
+Multi-layer writer
+
+```yaml
+name: napari-svg
+contributions:
+  commands:
+    - id: napari-svg.svg_writer
+      python_name: napari-svg.hook_implementations:writer
+      title: Save layers as SVG
+  writers:
+    - command: napari_svg.svg_writer
+      layer_types: ["image*", "labels*", "points*", "shapes*", "vectors*"]
+      filename_extensions: [".svg"]
+```
+
+### Layer type constraints
+
+Given a set of layers, compatible writer plugins are selected based their
+_layer type constraints_.
+
+A writer plugin can declare that it will write between _m_ and _n_ layers of a
+specific type where _0≤m≤n_.
+
+For example:
+
+```
+    image      Write exactly 1 image layer.
+    image?     Write 0 or 1 image layers.
+    image+     Write 1 or more image layers.
+    image*     Write 0 or more image layers.
+    image{k}   Write exactly k image layers.
+    image{m,n} Write between m and n layers (inclusive range). Must have m<=n.
+```
+
+When a type is not present in the list of constraints, that corresponds to a
+writer that is not compatible with that type. For example, a writer declaring:
+
+```
+    layer_types=["image+", "points*"]
+```
+
+would not be selected when trying to write an `image` and a `vector` layer
+because the above only works for cases with 0 `vector` layers.
+
+Note that just because a writer declares compatibility with a layer type does
+not mean it actually writes that type. In the example above, the writer might
+accept a set of layers containing `image`s and `point`s, but the write command
+might just ignore the `point` layers. The writer must return `None` for
+unwritten layers.
+
+### Calling convention
```

Comment:
we use the term function signature most other places.  any reason not to keep that here?

Reply:
I use "calling convention" to refer to both the function signature and the requirements around when it's called. 

For example, single and multi layer writers have different function signatures. Their behavior relative to one another is described by a calling convention.

Another example is the "autogenerate" flag modifying function widgets. The calling convention describes that, in the context where the function is invoked, it will be wrapped to produce a widget.